gateway-service-impl/src/main/java/org/hypertrace/gateway/service/explore/RequestHandler.java

```diff
@@ -532,6 +538,73 @@ private Map<String, AttributeMetadata> remapAttributeMetadataByResultName(
         attributeMetadataByIdMap);
   }
 
+  private org.hypertrace.gateway.service.v1.common.Filter createEntityIdAttributeFilter(
+      List<String> entityIdAttributes,
+      org.hypertrace.gateway.service.v1.common.Operator operator,
+      List<String> entityIds) {
+    if (entityIdAttributes.size() != 1) {
+      throw new RuntimeException("Cannot have more than one id attribute for an entity");
+    }
+
+    if (entityIds.isEmpty()) {
+      // Having empty entity ids is valid filter because this means that
+      // EDS source has filtered out all the entities and the result should
+      // be empty. But QS doesn't recognize empty IN filter as valid filter
+      // and ignoring this filter will generate wrong results.
+      //
+      // We could have return empty but clients expects response to contain
+      // information as per the selections they have sent. So we are converting
+      // empty entity Id filter into a false filter (id != null && id == null)
+      // so that results are always empty but must contain the valid selections.
+      return org.hypertrace.gateway.service.v1.common.Filter.newBuilder()
+          .setOperator(Operator.AND)
+          .addChildFilter(
+              org.hypertrace.gateway.service.v1.common.Filter.newBuilder()
+                  .setLhs(buildStringExpression(entityIdAttributes.get(0)))
+                  .setOperator(Operator.NEQ)
+                  .setRhs(buildStringExpression(null))
```

Comment:
we have a null literal expression for QS I thought? I guess it doesn't really matter, unless pinot optimizes for it

Reply:
We have it but at the query service layer. This is gateway service contract. I haven't tested it yet. Was waiting on whether this change is making sense or not

Reply:
Verified parsing of string null values is being handled by query service